python/pyarrow/array.pxi

```diff
@@ -161,7 +161,9 @@ def array(object obj, type=None, mask=None, size=None, from_pandas=None,
     else:
         c_from_pandas = from_pandas
 
-    if _is_array_like(obj):
+    if hasattr(obj, '__arrow_array__'):
+        return obj.__arrow_array__(type=type)
```

Comment:
Should we raise if non-default `size` and `mask` arguments are passed here?

Reply:
Yes, that sounds as a good idea.